quantecon/dle.py

```diff
@@ -0,0 +1,234 @@
+
+        self.H = np.hstack((self.llambda,self.pih.dot(uc).dot(phiin).dot(self.gamma),self.pih.dot(uc).dot(phiin).dot(self.ud)-self.ub,-self.pih.dot(uc).dot(phiin).dot(self.phii)))
+        self.G = ug.dot(phiin).dot(np.hstack((np.zeros((self.nd,self.nh)),self.gamma,self.ud,-self.phii)))
+        self.S = (self.G.T.dot(self.G) + self.H.T.dot(self.H))/2
+
+        self.nx = self.nh+self.nk+self.nz
+        self.n = self.ni+self.nh+self.nk+self.nz
+
+        self.R = self.S[0:self.nx,0:self.nx]
+        self.W = self.S[self.nx:self.n,0:self.nx]
+        self.Q = self.S[self.nx:self.n,self.nx:self.n]
+
+        # === Use quantecon's LQ code to solve our LQ problem === #
+
+        lq = LQ(self.Q, self.R, self.A, self.B, self.C, N=self.W, beta=self.beta)
+
+        self.P, self.F, self.d = lq.stationary_values()
+
+        # === Construct output matrices for our economy using the solution to the LQ problem === #
+
+        self.A0 = self.A - self.B.dot(self.F)
+
+        self.Sh = self.A0[0:self.nh,0:self.nx]
+        self.Sk = self.A0[self.nh:self.nh+self.nk,0:self.nx]
+        self.Sk1 = np.hstack((np.zeros((self.nk,self.nh)),np.eye(self.nk),np.zeros((self.nk,self.nz))))
+        self.Si = -self.F
+        self.Sd = np.hstack((np.zeros((self.nd,self.nh+self.nk)),self.ud))
+        self.Sb = np.hstack((np.zeros((self.nb,self.nh+self.nk)),self.ub))
+        self.Sc = uc.dot(phiin).dot(-self.phii.dot(self.Si) + self.gamma.dot(self.Sk1) + self.Sd)
+        self.Sg = ug.dot(phiin).dot(-self.phii.dot(self.Si) + self.gamma.dot(self.Sk1) + self.Sd)
+        self.Ss = self.llambda.dot(np.hstack((np.eye(self.nh),np.zeros((self.nh,self.nk+self.nz))))) + self.pih.dot(self.Sc)
+
+        # ===  Calculate eigenvalues of A0 === #
+        self.A110 = self.A0[0:self.nh+self.nk,0:self.nh+self.nk]
+        self.endo = np.linalg.eigvals(self.A110)
+        self.exo = np.linalg.eigvals(self.a22)
+
+        # === Construct matrices for Lagrange Multipliers === #
+
+        self.Mk = -2*np.asscalar(self.beta)*(np.hstack((np.zeros((self.nk,self.nh)),np.eye(self.nk),np.zeros((self.nk,self.nz))))).dot(self.P).dot(self.A0)
+        self.Mh = -2*np.asscalar(self.beta)*(np.hstack((np.eye(self.nh),np.zeros((self.nh,self.nk)),np.zeros((self.nh,self.nz))))).dot(self.P).dot(self.A0)
+        self.Ms = -(self.Sb - self.Ss)
+        self.Md = -(np.linalg.inv(np.vstack((self.phic.T,self.phig.T))).dot(np.vstack((self.thetah.T.dot(self.Mh) + self.pih.T.dot(self.Ms),-self.Sg))))
+        self.Mc = -(self.thetah.T.dot(self.Mh) + self.pih.T.dot(self.Ms))
+        self.Mi = -(self.thetak.T.dot(self.Mk))
+        #Note to check: I had to put a minus in front of everything to get same as the Matlab code
+        
+    def compute_steadystate(self, nnc=2): #nnc is the position of the constant in the state vector
+        """ Find non-stochastic steady-state of the economy """
+        zx = Matrix(np.eye(self.A0.shape[0])-self.A0)
+        self.zz = zx.nullspace()
```

Comment:
Can this be replaced with `quantecon.nullspace`? Then the import of `Matrix` from `sympy` would be removed.

Reply:
@oyamad I am in favour of this. Can you let me know what I need to change?

Reply:
``self.zz = qe.nullspace(A0)``

Reply:
@mmcky What I had in mind is

```py
zx = np.eye(self.A0.shape[0]) - self.A0
self.zz = qe.nullspace(zx)
```

Maybe the following is better?

```py
zx = -self.A0
zx[np.diag_indices_from(zx)] += 1
self.zz = qe.nullspace(zx)
self.zz /= self.zz[nnc]
```

But please check whether this gives the same shape of `self.zz` as the original code.

Reply:
thanks @oyamad. I will make the update and then confirm.